### ISSUE TITLE
Question: best option for new development of accessible carousels?

### ISSUE THREAD

**Issue by davidscotson** [Accessible360/accessible-slick#33]
Hi,

I was wondering if you had any opinions on the best way to create accessible content of this kind when not trying to re-use existing libraries. You have this slick.js fork and some suggestions for other popular libraries that have existed for a while in your codepen, but it seems like the bottom line is that if given the choice you wouldn't use these approaches.

Is there another approach you'd recommend for people not already tied to an existing library to achieve similar goals and would there be any benefit in mapping out an upgrade path for people currently using slick (and/or other libraries)? For example, do any of the new CSS scroll-snap properties help from an accessibility point of view?

**Comment by jasonwebb** [comment]
Hey David, good question! There are so many different ways to build carousels, each with their own pros and cons, that I don't think I can really say which is "best" in a generalized sense. Different use cases benefit from different techniques, and different dev teams have different stack preferences, so while one team might prefer a JavaScript-heavy implementation, another might prefer to use more CSS.

IMHO, from an accessibility perspective, the technical implementation details are less important than the end result for users. I recently wrote an article outlining the key elements and behaviors of that _any_ accessible carousel should have, regardless of the technologies used to implement it: https://dev.to/jasonwebb/how-to-build-a-more-accessible-carousel-or-slider-35lp

> Is there another approach you'd recommend for people not already tied to an existing library to achieve similar goals [...]?

My best recommendation for carousels is to not use them! If the decision is coming from a design team or management, there could be a good opportunity for education to help them understand the accessibility and usability issues that come with carousels. So as far as a recommended upgrade path for existing Slick users, I would probably lean more towards providing resources to help folks move away from carousels in general, rather than move on to another library that has the same inherent usability issues.

Side note: as far as I can tell, the [CSS scroll snap](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Scroll_Snap) properties really have no impact for accessibility, but they might have some benefits on the animation side since certain CSS properties can be GPU accelerated, IIRC. But regardless of whether you're using CSS scroll snap, jQuery's `animate()`, or a motion framework like GSAP, there are still things that need to be done that aren't handled by these techniques, like ensuring non-visible slides are unreachable for all users, and dynamically managing ARIA attributes. Unfortunately there is really no silver bullet here from a technological standpoint, other than building or configuring carousels that embrace accessibility as a core requirement and baking it in.

**Comment by davidscotson** [comment]
Hi Jason,

thanks for the comprehensive answer.

I'm still wondering though if there's some subset of "carousel" functionality that can be done more simply, beautifully and accessibly if you're not tied to the classic full-screen carousel design. For example, one of the default Slick examples, that you've reproduced is the responsive display (https://accessible360.github.io/accessible-slick/#responsive-display). In theory this could have 50 or more items in it and be used to focus on a tiny subset, but in the example, and I think in many real-world uses, a more reasonable number like the 8 in the demo might be used, and the main purpose is to make better use of space on mobile devices. e.g for large screens the user would see all of the items and the carousel JS would only need to kick in at smaller sizes, to provide a swipeable interface.

Would all those 8 items being always "visible" to a screen-reader as a list be a bad thing? This seems similar to making HTML tables responsive by wrapping them in a div and using overflow so they don't blow out the whole page if they're too wide. I'm assuming/hoping that the browser has enough information in the markup to present that table to various users without either a) abitrarily restricting access to parts of the data just because the screen size isn't enough, b) needing complex JS to ensure that the accessible tools know exactly which parts of the table are currently visible on an arbitrary size of screen and keep users updated on that as it changes.

There's also another re-implementation of Slick.js called Glide.js that simplifies things by dropping the ability to infinitely rotate, I can't actually find the place where they explain this decision, but basically, I think it drastically simplifies things as the simple list markup can be used and the rest is done mostly with CSS, and therefore doesn't need everything duplicated like slick.js and similar libraries often do to achieve their effects (and therefore do by default even for simple cases).

The name Carousel (as in the real life objects with horses on poles) kind of implies that this circular rotation is a fundamental requirement, but it is just an option in Slick.js currently. Many uses of the library will not actually rotate infinitely, or automatically and there's lots of people complaining about flexbox and slick.js not working well together.

I'm assuming for those situations a simpler approach overall might be possible and the "carousel" treated more like a simple list from accessability perspective until some certain options get used (like infinite rotation, auto movement, or more than 7-9 items maybe), at which time it needs the full current accessible slick.js implementation to properly reflect the experience that is being delivered by the Slick.js. But for many uses it seems to me (perhaps naively?) that the most accessible form would just be a list rather than try to reflect the swipability intended for improving experience and visuals on ipads and mobile devices?

I can't really say no to carousels even if I try to steer people away from them, but I'd like to be able to use a library for carousel-type things that does the best possible thing automatically for common use cases without having to hand-write things every time, and I feel like there's a gap in the market here to both increase accessibility, but also to simplify things for developers, even if it was specifically designed not to rotate or auto-play by default (or at least I'm not aware of an existing library that fits this niche).

**Comment by jasonwebb** [comment]
> Would all those 8 items being always "visible" to a screen-reader as a list be a bad thing?

I've chatted about this option a few times with my team over at @Accessible360 (which includes native screen reader and keyboard users), and each time I've found that they feel strongly that this would not be an equitable experience because it effectively forces keyboard and screen reader users to navigate through ALL of the carousel content when mouse users are free to choose which slides to look at.

In theory you might be able to argue that this is defensible as "accessible" under a strict interpretation of WCAG, but my team tends to prefer solutions that are unambiguously accessible and _equitable_ (as in, as similar as possible) whenever possible.

In my experience, this solution also turns out to be difficult, maybe even impossible to implement, because the non-visible content would need to be scrolled into view when a keyboard or screen reader user reaches it, since many keyboard and screen reader users have some degree of sight and are using these tools as aids. For example, elderly users, certain cognitively impaired (particularly with conditions that affect reading or language comprehension), and people with limited vision or other visual impairments (like cataracts) could be using a screen reader along with whatever level of sight they have.

If the slide track is implemented as a horizontally-overflowing container, then it could natively scroll horizontally whenever keyboard focus lands on an actionable element. But a screen reader user who is navigating the content with their virtual cursor would **not** be firing any events or forcing the browser to scroll when navigating the non-focusable elements (text, images, headings, etc), so browsers and custom JS would not know when content needs to be scrolled into view. So once they reach that non-visible content, they'd keep hearing stuff being read out to them, but they would likely have no idea where it is or how much of it is left.

> This seems similar to making HTML tables responsive by wrapping them in a div and using overflow [...]

Tables convey relational information where things like row headers and column headers may be necessary for understanding, whereas carousels slides are generally independent of each other and can be understood individually. WCAG 2.1's [1.4.10 guideline](https://www.w3.org/WAI/WCAG21/Understanding/reflow.html) actually specifically exempts data tables from reflow requirements, allowing them to be scrollable in two dimensions, probably because tables can't be understood properly when their data cells are dynamically shown/hidden.
